### PR TITLE
[AIRFLOW-5379] Add Google Search Ads 360 operators

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5379
### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/gcp/example_dags/example_search_ads.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    # [START howto_search_ads_generate_report_operator]
+    generate_report = GoogleSearchAdsInsertReportOperator(
+        report=REPORT, task_id="generate_report"
+    )
+    report_id = "{{ task_instance.xcom_pull('generate_report', key='report_id') }}"
+    # [END howto_search_ads_generate_report_operator]
+
+    # [START howto_search_ads_get_report_operator]
+    wait_for_report = GoogleSearchAdsReportSensor(
+        report_id=report_id, task_id="wait_for_report"
```
Comment (mik-laj) [inline airflow/gcp/example_dags/example_search_ads.py]:
```suggestion
        report_id="{{ task_instance.xcom_pull('generate_report', key='report_id') }}", task_id="wait_for_report"
```
I think it's worth including in one snippet of code - `[START/END]`

File: airflow/gcp/operators/search_ads.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+        gcs_hook = GoogleCloudStorageHook(
+            gcp_conn_id=self.gcp_conn_id, delegate_to=self.delegate_to
+        )
+
+        self.log.info("Downloading Search Ads report %s", self.report_id)
+
+        # Resolve report fragments
+        response = hook.get(report_id=self.report_id)
+        fragments = range(len(response["files"]))
```
Comment (mik-laj) [inline airflow/gcp/operators/search_ads.py]:
```suggestion
        fragments_count =len(response["files"])
```
This synax is confusing.

**Comment by mik-laj** [comment]
Can you ad this service to `integration.rst` file?

File: airflow/gcp/operators/search_ads.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+        # Download chunks of report's data
+        with NamedTemporaryFile() as temp_file:
+            for i in fragments:
```
Comment (mik-laj) [inline airflow/gcp/operators/search_ads.py]:
```suggestion
            for i in range(fragments_count):
```

File: airflow/gcp/operators/search_ads.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        self.report_id = report_id
+        self.chunk_size = chunk_size
+        self.gzip = gzip
+        self.bucket_name = self._set_bucket_name(bucket_name)
+        name = report_name or report_id
+        self.report_name = self._set_report_name(name)
+
+    @staticmethod
+    def _set_report_name(name: str) -> str:
+        return name if name.endswith(".csv") else name + ".csv"
```
Comment (efolgar) [inline airflow/gcp/operators/search_ads.py]: Can we add the ".gz" extension for compressed files?

File: airflow/gcp/sensors/search_ads.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any. For this to work, the service accountmaking the
+        request must have  domain-wide delegation enabled.
+    :type delegate_to: str
+    """
+
+    template_fields = ("report_id",)
+
+    @apply_defaults
+    def __init__(
```
Comment (efolgar) [inline airflow/gcp/sensors/search_ads.py]: Can we default the mode to "reschedule" and set the default poke interval to 5 minutes to better support larger reports?

File: airflow/gcp/operators/search_ads.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        )
+
+        gcs_hook = GoogleCloudStorageHook(
+            gcp_conn_id=self.gcp_conn_id, delegate_to=self.delegate_to
+        )
+
+        self.log.info("Downloading Search Ads report %s", self.report_id)
+
+        # Resolve report fragments
+        response = hook.get(report_id=self.report_id)
```
Comment (efolgar) [inline airflow/gcp/operators/search_ads.py]: Should we validate that the report is ready for downloading `response["isReportReady"]` before downloading?

**Review by efolgar — CHANGES_REQUESTED** [review]
Reviewed the hook, operators, and sensor for SA360.

**Comment by mik-laj** [comment]
@nuclearpinguin Can you add this operators to `operators-and-hooks-ref.rst` file?

**Comment by potiuk** [comment]
Approved but we need to move to the new providers/google packages.

**Review by mik-laj — APPROVED** [review]

**Pull request merged**